[Dino-Nuggies] iP

Dino-Nuggies opened the pull request:
# Lemon 🍋

> “Our brains are filled with ooey gooey substances" - Me

Lemon helps organise your thoughts to help you remember tasks. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use
- $${\color{red}Very Cringe}$$

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks using the predefined commands.
4. let it manage your tasks for you 😉

And it is FREE!

Features:

- [x] Managing tasks
- [ ] Reminders (coming soon)
- [ ] Bake a cake?

---

It uses `Java`!

``` Java
public class Main {
    public static void main(String[] args) {
        Lemon lemon = new Lemon();
        lemon.runLemon();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Dino-Nuggies committed:
Remove Duke in place of Lemon

Removed Duke
Added Lemon

Dino-Nuggies committed:
Added echo to Lemon

Added Scanner to lemon to be able to read input and echo it out

Dino-Nuggies committed:
Added list to lemon

Added a list of task lemon will track
Assuming less than 100 tasks

Dino-Nuggies committed:
Added lemon mark as done

Added the ability to mark task as done and unmark\
Added a task class
Fixed indentation bugs

Dino-Nuggies committed:
Added ToDos, Events, Deadlines to lemon

Added ToDos, Events, Deadlines classes
Added for lemon to be able to add special events
- Default Todo task type

Dino-Nuggies committed:
Add exception handling

Added exceptions for invalid commands, empty descriptions  and other wrong input handling
Added extra information being printed out for these errors
Fixed some output string extra spaces

Dino-Nuggies committed:
Add deleteTask

Add the ability to delete a task
Cleaned up code

Dino-Nuggies committed:
Add enum and switch case

Replaced the if else with enum and switch cases

shardhrv commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+public class Deadline extends Task {
+    protected String by;
```
You can consider turning the date for the by to be of `LocalDateTime` and add in checks for the user to format the code as such so that you can do queries via the date.

shardhrv commented on src/main/java/DescriptionException.java:
```diff
@@ -0,0 +1,5 @@
+public class DescriptionException extends Exception {
```
Could add JavaDoc to explain what this exception is for.

shardhrv commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,15 @@
+public class Event extends Task{
+    protected String from, to;
```
Similarly could use `LocalDateTime` here

shardhrv commented on src/main/java/InvalidCommandException.java:
```diff
@@ -0,0 +1,5 @@
+public class InvalidCommandException extends Exception{
```
Logical class name for the exception makes sense to me!

Zeepheru replied:
Yup!

shardhrv commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,156 @@
+import java.util.ArrayList;
+
+public class Lemon {
+    /*** Messages ***/
+    String logoMsg = "____________________________________________________________\n"
+            + " Hello! I'm Lemon\n"
+            + " What can I do for you?\n";
+    String endMsg = " Bye. Hope to see you again soon!\n"
+            + "____________________________________________________________\n";
+    String barMsg = "____________________________________________________________";
+    String emptyListMsg = " Sowwy, theres currently no tasks in your list.\n Ill be happy to add some for you OwO!";
+    String listMsg = " Here are the tasks in your list:";
+    String markMsg = " Nice! I've marked this task as done:";
+    String unmarkMsg = " OK, I've marked this task as not done yet:";
+    String addTaskMsg = " Got it. I've added this task:";
+    String deleteTaskMsg = " Noted. I've removed this task:";
```
Perhaps this could be refactored into a separate UI class.

shardhrv commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,156 @@
+    ArrayList<Task> tasks = new ArrayList<>();
+    int numTasks = 0;
+
+    enum Commands {
+        BYE,
+        LIST,
+        MARK,
+        UNMARK,
+        TODO,
+        EVENT,
+        DEADLINE,
+        DELETE
+    }
```
Good use of enums in your code!

shardhrv commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,156 @@
+        DELETE
+    }
+    
+    public static void main(String[] args) {
```
Could refactor the main function so that your chatbot has a singular run function that controls when it runs.

shardhrv commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,156 @@
+                    System.out.println(lemon.markMsg);
+                    lemon.tasks.get(next - 1).markDone();
+
+                    System.out.println("   " + lemon.tasks.get(next - 1).toString());
+                    break;
+                }
+                case UNMARK: {
+                    int next = scan.nextInt();
+                    if (next > lemon.numTasks || next <= 0) {
+                        throw new InvalidCommandException(" OOPS!!! Please select a valid task");
+                    }
+
+                    System.out.println(lemon.unmarkMsg);
+                    lemon.tasks.get(next - 1).unmarkDone();
+
+                    System.out.println("   " + lemon.tasks.get(next - 1).toString());
+                    break;
+                }
+                case TODO: {
+                    String next = scan.nextLine();
+
+                    lemon.addNewTask(new Todo(next));
+                    break;
+                }
+                case DEADLINE: {
+                    String[] next = scan.nextLine().split("/by ");
+
+                    lemon.addNewTask(new Deadline(next[0], next[1]));
+                    break;
+                }
+                case EVENT: {
+                    String[] next = scan.nextLine().split("/from ");
+
+                    lemon.addNewTask(new Event(next[0], next[1]));
+                }
+                case DELETE: {
+                    int next = scan.nextInt();
+                    if (next > lemon.numTasks || next <= 0) {
+                        throw new InvalidCommandException(" OOPS!!! Please select a valid task");
+                    }
```
This logic could be refactored out into a class that would parse the inputs. Clear what each switch-case does so good job!

shardhrv commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+public abstract class Task {
```
Ah good idea to make the Task class abstract!

shardhrv reviewed (commented):
In summary, your code follows good naming conventions, contributing to its readability and maintainability. 

The main areas for improvement are in using more JavaDocs and also refactoring certain code into separate classes to reduce responsibility of the main chatbot class.

These changes will make your code even more intuitive and easier for others to understand.

Keep up the good job! :))

Dino-Nuggies committed:
Add file saving support

Added the possibility to save the file in a local text file

Dino-Nuggies committed:
Add Date support

Event and Deadline now stores dates in the format of
dd-MM-yyyy

Dino-Nuggies committed:
Make code more OOP

Dino-Nuggies committed:
Move files into relevant packages

Dino-Nuggies committed:
Add gradle

Dino-Nuggies committed:
Add gradle

Dino-Nuggies committed:
Add JavaDoc to public functions and classes

Zeepheru commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,22 @@
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + "(by: " +
+                by.format(DateTimeFormatter.ofPattern("MMM dd yyyy")) + ")";
```
Could consider using `String.format()` instead of concatenation for complex Strings?

For instance, this `toString()` here would be
```java
return String.format("[D]%s(by: %s)", 
        super.toString(), by.format(DateTimeFormatter.ofPattern("MMM dd yyyy"));
```

Zeepheru commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,286 @@
+                                LocalDate.parse(temp[3], DateTimeFormatter.ofPattern("yyyy-MM-dd")),
+                                LocalDate.parse(temp[4], DateTimeFormatter.ofPattern("yyyy-MM-dd")),
+                                Boolean.parseBoolean(temp[1])));
+                        break;
```
For Switch statements, cases should not be indented (following the style guide)
Aka:
```java
switch (temp[0]) {
case "T":
    addNewTask(...);
    // ...
case "D":
// ...
}
```

Might have to set this in your IDE formatting options, default styling normally has the cases indented.

_Actually you have another switch statement in your code that's formatted correctly haha_

Zeepheru commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,286 @@
+        } catch (IOException e) {
+            isInitialised = false;
+            System.out.print(" Im sowwy... Something went wrong, QwQ. Unable to create file.\n" +
+                    " I dont think i can do this anymore");
```
*I approve of this style of message* :3

Zeepheru commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,286 @@
+            String input = systemScanner.next().toUpperCase();
+            System.out.println(barMsg);
+            try {
+                Commands command = Commands.valueOf(input);
```
Neat way of translating user input to commands by matching with the Enum commands by using `.toUpperCase()` then `Commands.valueOf(input)`!

Zeepheru commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
Good to have the explainer comment.

Zeepheru commented on data/lemonSaves.txt:
```diff
@@ -0,0 +1,2 @@
+D|false| help me |2023-03-22
+E|false| murder me |2023-04-03|2023-04-03
```
Aww, those task names :c

---

Generally it's not a good idea to not push data files like this to a remote repository. Otherwise if you had actual tasks saved to the file, strangers like me can see your *potentially private* data :O

You can consider adding the following to the `.gitignore` file in your root directory:

```bash
# current content of .gitignore
# ...
/data/
```

So that when you add files to a git commit, git knows to ignore the directory `./data/` :D

Zeepheru commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,286 @@
+            try {
+                Commands command = Commands.valueOf(input);
+                switch (command) {
+                case BYE: {
+                    boolean isAbleToSave = saveTasks();
+                    if (isAbleToSave) {
+                        isRunning = false;
+                        System.out.println(endMsg);
+                    }
+                    break;
+                }
```
Correct indentation for case in this switch statement. 

Though, braces aren't required. It can just be:
```java
switch (command) {
case BYE:
    boolean isAbleToSave = ...
    // ...
    break;
```

Zeepheru reviewed (commented):
Code follows correct naming conventions, and generally follows the correct style following the code quality guidelines.

Keep up the good work ^^

Dino-Nuggies committed:
Add JUnit testing for Storage and TaskList

Dino-Nuggies committed:
Fix conflicts

Dino-Nuggies committed:
Add find

Use find as a way to find a task by searching for a keyword in the task description.

Dino-Nuggies committed:
Add JavaDoc

Dino-Nuggies committed:
Remove msg

Dino-Nuggies committed:
Add find

Add find again

Dino-Nuggies committed:
Add JavaDoc

for level 9 functions that was added

Dino-Nuggies committed:
Add gradle run

Dino-Nuggies committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Dino-Nuggies committed:
Fix gradle

Dino-Nuggies committed:
Add checkstyle requirement

Dino-Nuggies committed:
Add rudimentary javaFx support

todo: implement proper response for lemon and GUI to communicate

Dino-Nuggies committed:
Split commands into multiple command classes

lemon.executeLemon(): very long method that needs to be refactored

The extensiveness of the function can be hard to troubleshoot or add new commands. Additionally the error handling of the function is not done very clearly.

By refactoring and removing duplicated code can help clean up the code and improve code quality.

Change:
* parser class: Handle the switch cases for generating the right command to be executed
* command classes: Handle the implementation of the commands and necessary exception handling
* ui class: Handle printing of error messages
* Remove InvalidCommandException and DescriptionException classes as they are no longer used

Todo:
* More extensive testing of lemon.executeLemon() to make sure the new version does not have unexpected errors
* Add the necessary JavaDocs for newly implemented classes and methods once finalized

Dino-Nuggies committed:
Move files into packages

Dino-Nuggies committed:
Preparation for Fx integration

Dino-Nuggies committed:
Fix Javafx implementation

Fix lemon receiving and returning output when using GUI

Dino-Nuggies committed:
Add Assert to window within Response

Assert check for window to not be null

Dino-Nuggies committed:
Improve code quality

Added missing JavaDocs to newly added classes
Made sure the code comply with checkstyle

Dino-Nuggies committed:
Merge pull request #2 from Dino-Nuggies/branch-A-Assertions

Add Assert to window within Response

Dino-Nuggies committed:
Merge branch 'master' into branch-A-CodeQuality

Dino-Nuggies committed:
Revert "Merge branch 'master' into branch-A-CodeQuality"

This reverts commit e3ffcbabc9444fc4a1df75d0f39b5a104e593a2a, reversing
changes made to 81700b107fb9609399f6edaca6a9027435abb020.

Dino-Nuggies committed:
Fix merge conflicts

Dino-Nuggies committed:
Merge pull request #3 from Dino-Nuggies/branch-A-CodeQuality

Branch a code quality

Dino-Nuggies committed:
Add a help command for new user to navigate

This is related to the task under C-Help from the extension SE page

Dino-Nuggies committed:
Add better pictures for lemon

Dino-Nuggies committed:
Add more code quality

Made sure that the newly implemented help command comply with checkstyle

Dino-Nuggies committed:
Add project ReadMe and lemon Ui image

Dino-Nuggies committed:
Fix image not showing

Dino-Nuggies committed:
Update README.md
